File: wearshield-apple/WearShield/Bloom/Kotoba.swift
// Kotoba — glyph fold operators from konomigami
// Each glyph is a fold instruction on the bloom state

import Foundation

struct KotobaGlyph {
    let id: String
    let name: String
    let prime: Int?
    let ring: Int?
    let foldType: String
    let apply: (inout [Double]) -> Void
}

struct Kotoba {
    static let glyphs: [Character: KotobaGlyph] = [
        "●": KotobaGlyph(id: "ground", name: "GROUND", prime: 2, ring: 0, foldType: "identity") { b in },
        "〜": KotobaGlyph(id: "wave", name: "WAVE", prime: 3, ring: 1, foldType: "valley") { b in
            b[1] = min(5, b[1] + 0.5)
        },
        "┃": KotobaGlyph(id: "gate", name: "GATE", prime: 5, ring: 2, foldType: "mountain") { b in
            b[2] = min(5, b[2] + 0.5)
        },
        "♡": KotobaGlyph(id: "sink", name: "SINK", prime: 7, ring: 3, foldType: "sink") { b in
            b[3] = min(5, b[3] + 1)
        },
        "△": KotobaGlyph(id: "reverse", name: "REVERSE", prime: 11, ring: 4, foldType: "reverse") { b in
            b[4] = min(5, b[4] + 0.5)
        },
        "◐": KotobaGlyph(id: "petal", name: "PETAL", prime: 13, ring: 5, foldType: "petal") { b in
            b[5] = min(5, b[5] + 0.5)
        },
        "◯": KotobaGlyph(id: "collapse", name: "COLLAPSE", prime: 17, ring: 6, foldType: "waterbomb") { b in
            b[6] = min(5, b[6] + 1)
        },
        "火": KotobaGlyph(id: "fire", name: "FIRE", prime: nil, ring: nil, foldType: "double") { b in
            for i in 0..<7 { b[i] = min(5, b[i] * 1.3) }
        },
        "水": KotobaGlyph(id: "water", name: "WATER", prime: nil, ring: nil, foldType: "halve") { b in
            for i in 0..<7 { b[i] = max(1, b[i] * 0.7) }
        },
        "空": KotobaGlyph(id: "sky", name: "SKY", prime: nil, ring: nil, foldType: "noop") { _ in },
        "雷": KotobaGlyph(id: "thunder", name: "THUNDER", prime: nil, ring: nil, foldType: "snap") { b in
            for i in 0..<7 { b[i] = min(5, b[i] + 1) }
        },
        "響": KotobaGlyph(id: "echo", name: "ECHO", prime: nil, ring: nil, foldType: "repeat") { _ in },
        "華": KotobaGlyph(id: "bloom", name: "BLOOM", prime: nil, ring: nil, foldType: "unfold") { b in
            for i in 0..<7 { b[i] = max(1, b[i] - 0.3) }
        },
        "輪": KotobaGlyph(id: "wheel", name: "WHEEL", prime: nil, ring: nil, foldType: "rotate") { b in
            let last = b[6]; for i in stride(from: 6, through: 1, by: -1) { b[i] = b[i-1] }; b[0] = last
        },
        "工": KotobaGlyph(id: "craft", name: "CRAFT", prime: nil, ring: nil, foldType: "forge") { b in
            let avg = b.reduce(0, +) / 7; for i in 0..<7 { b[i] = avg }
        },
        "数": KotobaGlyph(id: "number", name: "NUMBER", prime: nil, ring: nil, foldType: "quantize") { b in
            for i in 0..<7 { b[i] = Double(Int(b[i] + 0.5)) }
        },
        "影": KotobaGlyph(id: "shadow", name: "SHADOW", prime: nil, ring: nil, foldType: "invert") { b in
            for i in 0..<7 { b[i] = 6 - b[i] }
        }
    ]

    static func apply(_ text: String, to bloom: inout [Double]) -> [Character] {
        var fired: [Character] = []
        for ch in text {
            if var g = glyphs[ch] {
                g.apply(&bloom)
                fired.append(ch)
            }
        }
        return fired
    }
}

File: js/kotoba.js
// Κ.KOTOBA — glyph state mutations
export const KOTO = {
  '●': { nm: 'ground', ring: 0, fn: s => { s.primary = 'bloomqr' } },
  '〜': { nm: 'wave', ring: 1, fn: s => { s.tint += 30 } },
  '┃': { nm: 'gate', ring: 2, fn: s => { s.depth = Math.min(12, s.depth + 1) } },
  '♡': { nm: 'heart', ring: 3, fn: s => { s.emo = 'love' } },
  '△': { nm: 'think', ring: 4, fn: s => { s.primary = 'shield' } },
  '◐': { nm: 'self', ring: 5, fn: s => { s.dens = Math.min(3, s.dens + .5) } },
  '◯': { nm: 'witness', ring: 6, fn: s => { s.primary = 'bloomqr' } },
  '火': { nm: 'fire', fn: s => { s.tint += 330; s.dens = Math.min(3, s.dens + .3) } },
  '水': { nm: 'water', fn: s => { s.tint += 210; s.depth = Math.max(1, s.depth - 1) } },
  '空': { nm: 'sky', fn: s => { s.depth = Math.min(12, s.depth + 2) } },
  '雷': { nm: 'thunder', fn: s => { s.dens = Math.min(3, s.dens + 1) } },
  '響': { nm: 'echo', fn: s => { s.tint += 60 } },
  '華': { nm: 'bloom', fn: s => { s.emo = 'bloom' } },
  '輪': { nm: 'wheel', fn: s => { s.primary = 'mahoraga' } },
  '工': { nm: 'craft', fn: s => { s.primary = 'factory' } },
  '数': { nm: 'number', fn: s => { s.depth = 7 } },
  '影': { nm: 'shadow', fn: s => { s.tint += 180; s.dens = Math.max(.3, s.dens - .3) } }
}

export function applyGlyphs(text, state) {
  let fired = []
  for (const ch of text) {
    if (KOTO[ch]) { KOTO[ch].fn(state); fired.push(ch) }
  }
  return fired
}
